### DISCUSSION TITLE
More polyphony and effects possible with more powerful Pis?

### DISCUSSION THREAD

**Discussion by Kirtai** [probonopd/MiniDexed#621] [Ideas]
I've noticed that work has been ongoing on running more tone generators on the newer Pis but it occurred to me that they should also be able to support more polyphony and built in effects. Any thoughts on this?

**Comment by diyelectromusic** [comment]
I'm not sure anyone has experimented really.

The polyphony is set in https://github.com/probonopd/MiniDexed/blob/main/src/config.h and depends on the version of the Pi.  16 is the default for anything apart from a V1/Zero, which is 8.  Synth_Dexed itself appears to be able to go up to 32 note polyphony according to this: https://codeberg.org/dcoredump/Synth_Dexed/src/branch/master/src/synth_dexed.h#L40 so it would be worth giving it a go on a Pi 4 at least, and almost certainly would be fine on a Pi 5 when that comes along.

I don't really know how the effects subsystem works if I'm honest.  I know there were some PRs to add extra, more complex effects, but I don't think it was ever finished (see https://github.com/probonopd/MiniDexed/pull/499).

I tend to work at the other end of things - seeing what I can get away with on the older Pis (and more recently, even the Pico!).

Kevin

**Reply by Kirtai** [reply]
I've been following your Pico work, it's very cool.

I'm just curious to see how far it can be pushed in _both_ directions; taking advantage of the biggest _and_ smallest Pi hardware available :)

**Comment by Banana71** [comment]
>  Any thoughts on this?

## Polyphony:
- If the polyphony was increased, the piano sounds would particularly benefit. The sound engine sometimes produces noise (quiet cracking) when the maximum polyphony is exceeded.

- Before doubling the number of TG's in an Rpi 4 I would rather increase the number of miniDexeds in my setup. With multiple RPis you are more flexible when it comes to effects, audio outputs and engine type. In addition, the Performance.ini remains as it is.

- At the lower end (RPi 1 & RPi Zero) it could perhaps think more along the lines of TX-802. E.g. 4 TG's with a maximum of 2x polyphony each, for drum sounds, raisers or other effect sounds. Whether this makes sense given the availability of the RPi Zero 2 remains to be seen.

Kevin @diyelectromusic , could you please compile an installable version for the RPi 4 in which the maximum polyphony is 32?

## Effects:

I have a specific opinion about the effects and it depends heavily on the way I use it. But basically I believe less is more.

- It would be nice if you could also freely parameterize the compressor. That would help with the drum sounds that are far too quiet.

- With the existing reverb there is nothing to complain about except for the fact that there is no WET/DRY control.

- Otherwise, I'm missing a parametric 3-band equalizer that could be used to filter out critical frequencies from the sum signal. FM synthesis in particular can generate strong low frequencies that could be dangerous for some music systems.

- A stereo delay would be a big help for the sound design.

It occurs to me that I actually want the same effects section, except for the chorus, as in the Roland D50. I prefer to create the chorus effect by layering, detuning and panning several sounds.

**Reply by diyelectromusic** [reply]
No promises, as this builds but is largely untested, but feel free to see if this works.  I think it has increased the polyphony for a Pi 4 but it isn't particularly easy to tell! :)

[MiniDexed_2024-03-10-4db0188](https://github.com/probonopd/MiniDexed/actions/runs/8224964103/artifacts/1313262227)

(from https://github.com/probonopd/MiniDexed/pull/623)

Kevin

**Reply by Kirtai** [reply]
Yes, one of the things I was thinking about re effects was the problem of using outboard effects with only one output on the MiniDexed.

Some kind of chorus might be nice for that string machine sound :)

I'll have to move my MD to my Pi4 and maybe try a Zero2 too.

**Comment by Banana71** [comment]
Thanks Kevin, it works wonderfully. It's not just ambient artists like Brian Eno who enjoy it. 256 Voice at the same time also looks very good on the data sheet and would make many manufacturers change the price tag.

But there is also a **small** disadvantage: the oscillators for sounds with OSC Key Sync=OFF only really oscillate freely from the 32nd note played. This means that after a performance change, the sound only sounds really fat and wide in the stereo image from the 32nd note played onwards.

The question now is what do we do with the knowledge?
Should we extend the test to the RPi 3 and Zero 2 W?
Should the RPi 4 always receive 32x polyphony or do we create another line in the minidexed.ini and make it configurable?

Thanks again to Kevin 🎈

**Reply by diyelectromusic** [reply]
Well I guess the issue is really - is there a downside to a default 32-note polyphony on a PI 4?  I guess we need to attempt to push it to see if that will start to impair performance or not.  If not, then I see no reason not to leave it as 32.

Pi 3, I suspect it ought to stay as is, but that is just a feeling.  I guess it could be pushed to an intermediate value - 20 or 24 say, but again without someone pushing the bounds of what might be possible its hard to say.

It might be possible to make it a configuration setting... but the way it is all initialised at the moment it can't be.  It would need a bit of work I think to be configurable.

**Reply by Banana71** [reply]
I don't see anything that speaks against running the RPi4 permanently with a polyphony of 32.
In my opinion, there is no need to configure Polyphony.

**Reply by Kirtai** [reply]
Pi3 might be worth increasing because there's the Zero2 which can be used for super compact versions like the eurorack elsewhere in the discussions and for easier installing inside keyboards. And it would be nice if we could use all of its capabilities too.

**Reply by Kirtai** [reply]
Oh, and I've never had a problem with the Pi *2* so far so the 3 should be capable of more.

**Comment by Banana71** [comment]
> Some kind of chorus might be nice for that string machine sound :)

@Kirtai, What sound do you mean exactly?

**Reply by Kirtai** [reply]
I was mostly thinking of the string ensemble sounds you could get from an ARP Solina or Quadra. If I can get it another way that's great :)

**Reply by Banana71** [reply]
@Kirtai, I could continue in this direction

[000024_PhaserStrings.zip](https://github.com/probonopd/MiniDexed/files/14564813/000024_PhaserStrings.zip)

second try, I just made a mistake while copying :-(

**Reply by Kirtai** [reply]
Ooh, that's really nice, thank you :smile:

**Reply by Banana71** [reply]
> Some kind of chorus might be nice for that string machine sound :)

Just for fun...

https://github.com/probonopd/MiniDexed/assets/104296205/b7fc8bed-d619-47ff-b97a-b2a00d7fbcd2

I recommend headphones to hear the chorus effect

**Reply by Kirtai** [reply]
I love these :smile:

**Reply by probonopd** [reply]
In ["DX Dreams: Capturing the Magic of the Yamaha DX7"](https://www.youtube.com/watch?v=PD_6Xvg9iMQ), Jim Daneker specifically mentions the Roland Dimension D chorus and AMS RMX16 reverb around 4:40.

Personally I also like the Juno 6 chorus.

Are there any suitably licensed C++ implementations?

**Reply by probonopd** [reply]
And then, there is the pull request
https://github.com/probonopd/MiniDexed/pull/499

Would it do what you need?

**Reply by Kirtai** [reply]
Oh, that looks *very* cool, though banana71's word is worth more than mine here :)

**Reply by Banana71** [reply]
Espen Kraft once introduced this effects device. Recently there has been a successor which is of course a bit more expensive :-(

https://www.youtube.com/watch?v=OtEqwcrgZ4w

> Oh, that looks very cool, though banana71's word is worth more than mine here :)

The great thing about this forum is that every word is heard and treated equally. I like how the users treat each other here.

**Comment by Banana71** [comment]
### What's going through my head at the moment:

1. The maximum polyphony is automatically set depending on the Raspberry Pi used. (As already implemented in the last version)

2. My favorite sound engine is "1:Modern".
The sound engine types "2:Mark I" and "3:OPL" produce background noise (hissing or crackling) in many situations. But the two engines can handle algorithms 4 +6 properly.

3. I would like the miniDexed to automatically always choose the best configuration for me. Automatically means the miniDexed basically selects the “Modern” sound engine. Except in the current performance there is a patch with the algorithm 4 or 6 then the sound engine  is switched to Type 2 (Mark I), and the maximum polyphony is also limited.
  
```
# Engine Type ( 1=Modern ; 2=Mark I ; 3=OPL; 4=Auto )
EngineType=4
``` 

**Questions that need to be checked beforehand:**
- Can the miniDexed change the maximum polyphony without a cold start?
- Can the miniDexed change the sound engine during operation?
- Can a performance handle different sound engines? Or do you have to choose a sound engine for performance?

**Why all this:**
So far I have programmed all performances for Soundengine Modern. But there are many sounds that can only be implemented with algorithms 4 + 6 (Hoover sound, trance and crazy sound effects).

What do you think?

**Reply by diyelectromusic** [reply]
I'll need to do some digging around in the code to consider the art of the possible then get back to you :)
